// graphile-build-pg/node8plus/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.constantCase = exports.camelCase = exports.upperFirst = exports.PgMutationPayloadEdgePlugin = exports.PgMutationProceduresPlugin = exports.PgMutationUpdateDeletePlugin = exports.PgMutationCreatePlugin = exports.PgConnectionTotalCount = exports.PageInfoStartEndCursor = exports.PgScalarFunctionConnectionPlugin = exports.PgNodeAliasPostGraphQL = exports.PgNodeAliasPostGraphile = exports.PgRowNode = exports.PgOrderByPrimaryKeyPlugin = exports.PgOrderAllColumnsPlugin = exports.PgQueryProceduresPlugin = exports.PgComputedColumnsPlugin = exports.PgRowByUniqueConstraint = exports.PgBackwardRelationPlugin = exports.PgForwardRelationPlugin = exports.PgColumnDeprecationPlugin = exports.PgColumnsPlugin = exports.PgAllRows = exports.PgConnectionArgCondition = exports.PgConnectionArgOrderByDefaultValue = exports.PgConnectionArgOrderBy = exports.PgConnectionArgFirstLastBeforeAfter = exports.PgTablesPlugin = exports.PgJWTPlugin = exports.PgTypesPlugin = exports.PgIntrospectionPlugin = exports.PgBasicsPlugin = exports.inflections = exports.defaultPlugins = exports.omit = exports.parseIdentifier = undefined;

var _graphileBuild = require("graphile-build");

Object.defineProperty(exports, "upperFirst", {
  enumerable: true,
  get: function () {
    return _graphileBuild.upperFirst;
  }
});
Object.defineProperty(exports, "camelCase", {
  enumerable: true,
  get: function () {
    return _graphileBuild.camelCase;
  }
});
Object.defineProperty(exports, "constantCase", {
  enumerable: true,
  get: function () {
    return _graphileBuild.constantCase;
  }
});

var _PgBasicsPlugin = require("./plugins/PgBasicsPlugin");

var _PgBasicsPlugin2 = _interopRequireDefault(_PgBasicsPlugin);

var _PgIntrospectionPlugin = require("./plugins/PgIntrospectionPlugin");

var _PgIntrospectionPlugin2 = _interopRequireDefault(_PgIntrospectionPlugin);

var _PgTypesPlugin = require("./plugins/PgTypesPlugin");

var _PgTypesPlugin2 = _interopRequireDefault(_PgTypesPlugin);

var _PgJWTPlugin = require("./plugins/PgJWTPlugin");

var _PgJWTPlugin2 = _interopRequireDefault(_PgJWTPlugin);

var _PgTablesPlugin = require("./plugins/PgTablesPlugin");

var _PgTablesPlugin2 = _interopRequireDefault(_PgTablesPlugin);

var _PgConnectionArgFirstLastBeforeAfter = require("./plugins/PgConnectionArgFirstLastBeforeAfter");

var _PgConnectionArgFirstLastBeforeAfter2 = _interopRequireDefault(_PgConnectionArgFirstLastBeforeAfter);

var _PgConnectionArgOrderBy = require("./plugins/PgConnectionArgOrderBy");

var _PgConnectionArgOrderBy2 = _interopRequireDefault(_PgConnectionArgOrderBy);

var _PgConnectionArgOrderByDefaultValue = require("./plugins/PgConnectionArgOrderByDefaultValue");

var _PgConnectionArgOrderByDefaultValue2 = _interopRequireDefault(_PgConnectionArgOrderByDefaultValue);

var _PgConnectionArgCondition = require("./plugins/PgConnectionArgCondition");

var _PgConnectionArgCondition2 = _interopRequireDefault(_PgConnectionArgCondition);

var _PgAllRows = require("./plugins/PgAllRows");

var _PgAllRows2 = _interopRequireDefault(_PgAllRows);

var _PgColumnsPlugin = require("./plugins/PgColumnsPlugin");

var _PgColumnsPlugin2 = _interopRequireDefault(_PgColumnsPlugin);

var _PgColumnDeprecationPlugin = require("./plugins/PgColumnDeprecationPlugin");

var _PgColumnDeprecationPlugin2 = _interopRequireDefault(_PgColumnDeprecationPlugin);

var _PgForwardRelationPlugin = require("./plugins/PgForwardRelationPlugin");

var _PgForwardRelationPlugin2 = _interopRequireDefault(_PgForwardRelationPlugin);

var _PgBackwardRelationPlugin = require("./plugins/PgBackwardRelationPlugin");

var _PgBackwardRelationPlugin2 = _interopRequireDefault(_PgBackwardRelationPlugin);

var _PgRowByUniqueConstraint = require("./plugins/PgRowByUniqueConstraint");

var _PgRowByUniqueConstraint2 = _interopRequireDefault(_PgRowByUniqueConstraint);

var _PgComputedColumnsPlugin = require("./plugins/PgComputedColumnsPlugin");

var _PgComputedColumnsPlugin2 = _interopRequireDefault(_PgComputedColumnsPlugin);

var _PgQueryProceduresPlugin = require("./plugins/PgQueryProceduresPlugin");

var _PgQueryProceduresPlugin2 = _interopRequireDefault(_PgQueryProceduresPlugin);

var _PgOrderAllColumnsPlugin = require("./plugins/PgOrderAllColumnsPlugin");

var _PgOrderAllColumnsPlugin2 = _interopRequireDefault(_PgOrderAllColumnsPlugin);

var _PgOrderByPrimaryKeyPlugin = require("./plugins/PgOrderByPrimaryKeyPlugin");

var _PgOrderByPrimaryKeyPlugin2 = _interopRequireDefault(_PgOrderByPrimaryKeyPlugin);

var _PgRowNode = require("./plugins/PgRowNode");

var _PgRowNode2 = _interopRequireDefault(_PgRowNode);

var _PgNodeAliasPostGraphile = require("./plugins/PgNodeAliasPostGraphile");

var _PgNodeAliasPostGraphile2 = _interopRequireDefault(_PgNodeAliasPostGraphile);

var _PgScalarFunctionConnectionPlugin = require("./plugins/PgScalarFunctionConnectionPlugin");

var _PgScalarFunctionConnectionPlugin2 = _interopRequireDefault(_PgScalarFunctionConnectionPlugin);

var _PageInfoStartEndCursor = require("./plugins/PageInfoStartEndCursor");

var _PageInfoStartEndCursor2 = _interopRequireDefault(_PageInfoStartEndCursor);

var _PgConnectionTotalCount = require("./plugins/PgConnectionTotalCount");

var _PgConnectionTotalCount2 = _interopRequireDefault(_PgConnectionTotalCount);

var _PgMutationCreatePlugin = require("./plugins/PgMutationCreatePlugin");

var _PgMutationCreatePlugin2 = _interopRequireDefault(_PgMutationCreatePlugin);

var _PgMutationUpdateDeletePlugin = require("./plugins/PgMutationUpdateDeletePlugin");

var _PgMutationUpdateDeletePlugin2 = _interopRequireDefault(_PgMutationUpdateDeletePlugin);

var _PgMutationProceduresPlugin = require("./plugins/PgMutationProceduresPlugin");

var _PgMutationProceduresPlugin2 = _interopRequireDefault(_PgMutationProceduresPlugin);

var _PgMutationPayloadEdgePlugin = require("./plugins/PgMutationPayloadEdgePlugin");

var _PgMutationPayloadEdgePlugin2 = _interopRequireDefault(_PgMutationPayloadEdgePlugin);

var _inflections = require("./inflections");

var inflections = _interopRequireWildcard(_inflections);

var _parseIdentifier = require("./parseIdentifier");

var _parseIdentifier2 = _interopRequireDefault(_parseIdentifier);

var _omit = require("./omit");

var _omit2 = _interopRequireDefault(_omit);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.parseIdentifier = _parseIdentifier2.default;
exports.omit = _omit2.default;

// Mutations

const defaultPlugins = exports.defaultPlugins = [_PgBasicsPlugin2.default, _PgIntrospectionPlugin2.default, _PgTypesPlugin2.default, _PgJWTPlugin2.default, _PgTablesPlugin2.default, _PgConnectionArgFirstLastBeforeAfter2.default, _PgConnectionArgOrderBy2.default, _PgConnectionArgOrderByDefaultValue2.default, _PgConnectionArgCondition2.default, _PgAllRows2.default, _PgColumnsPlugin2.default, _PgColumnDeprecationPlugin2.default, _PgForwardRelationPlugin2.default, _PgBackwardRelationPlugin2.default, _PgRowByUniqueConstraint2.default, _PgComputedColumnsPlugin2.default, _PgQueryProceduresPlugin2.default, _PgOrderAllColumnsPlugin2.default, _PgOrderByPrimaryKeyPlugin2.default, _PgRowNode2.default, _PgNodeAliasPostGraphile2.default, _PgScalarFunctionConnectionPlugin2.default, // For PostGraphile compatibility
_PageInfoStartEndCursor2.default, // For PostGraphile compatibility
_PgConnectionTotalCount2.default,

// Mutations
_PgMutationCreatePlugin2.default, _PgMutationUpdateDeletePlugin2.default, _PgMutationProceduresPlugin2.default, _PgMutationPayloadEdgePlugin2.default];

exports.inflections = inflections;
exports.PgBasicsPlugin = _PgBasicsPlugin2.default;
exports.PgIntrospectionPlugin = _PgIntrospectionPlugin2.default;
exports.PgTypesPlugin = _PgTypesPlugin2.default;
exports.PgJWTPlugin = _PgJWTPlugin2.default;
exports.PgTablesPlugin = _PgTablesPlugin2.default;
exports.PgConnectionArgFirstLastBeforeAfter = _PgConnectionArgFirstLastBeforeAfter2.default;
exports.PgConnectionArgOrderBy = _PgConnectionArgOrderBy2.default;
exports.PgConnectionArgOrderByDefaultValue = _PgConnectionArgOrderByDefaultValue2.default;
exports.PgConnectionArgCondition = _PgConnectionArgCondition2.default;
exports.PgAllRows = _PgAllRows2.default;
exports.PgColumnsPlugin = _PgColumnsPlugin2.default;
exports.PgColumnDeprecationPlugin = _PgColumnDeprecationPlugin2.default;
exports.PgForwardRelationPlugin = _PgForwardRelationPlugin2.default;
exports.PgBackwardRelationPlugin = _PgBackwardRelationPlugin2.default;
exports.PgRowByUniqueConstraint = _PgRowByUniqueConstraint2.default;
exports.PgComputedColumnsPlugin = _PgComputedColumnsPlugin2.default;
exports.PgQueryProceduresPlugin = _PgQueryProceduresPlugin2.default;
exports.PgOrderAllColumnsPlugin = _PgOrderAllColumnsPlugin2.default;
exports.PgOrderByPrimaryKeyPlugin = _PgOrderByPrimaryKeyPlugin2.default;
exports.PgRowNode = _PgRowNode2.default;
exports.PgNodeAliasPostGraphile = _PgNodeAliasPostGraphile2.default;
exports.PgNodeAliasPostGraphQL = _PgNodeAliasPostGraphile2.default;
exports.PgScalarFunctionConnectionPlugin = _PgScalarFunctionConnectionPlugin2.default;
exports.PageInfoStartEndCursor = _PageInfoStartEndCursor2.default;
exports.PgConnectionTotalCount = _PgConnectionTotalCount2.default;
exports.PgMutationCreatePlugin = _PgMutationCreatePlugin2.default;
exports.PgMutationUpdateDeletePlugin = _PgMutationUpdateDeletePlugin2.default;
exports.PgMutationProceduresPlugin = _PgMutationProceduresPlugin2.default;
exports.PgMutationPayloadEdgePlugin = _PgMutationPayloadEdgePlugin2.default;
